diff --git a/arch/arm/dts/Makefile b/arch/arm/dts/Makefile
index bbe1887623..9a5d6b2347 100644
--- a/arch/arm/dts/Makefile
+++ b/arch/arm/dts/Makefile
@@ -546,6 +546,7 @@ dtb-$(CONFIG_MACH_SUN8I_R40) += \
 	sun8i-r40-bananapi-m2-ultra.dtb \
 	sun8i-v40-bananapi-m2-berry.dtb
 dtb-$(CONFIG_MACH_SUN8I_V3S) += \
+	sun8i-v3s-prince.dtb \
 	sun8i-v3s-licheepi-zero.dtb 
 dtb-$(CONFIG_MACH_SUN8I_S3) += \
 	sun8i-s3-generic.dtb
