diff --git a/arch/arm/mach-sunxi/board.c b/arch/arm/mach-sunxi/board.c
index 35c4293fd6..37c394457d 100644
--- a/arch/arm/mach-sunxi/board.c
+++ b/arch/arm/mach-sunxi/board.c
@@ -226,7 +226,7 @@ void s_init(void)
 #if defined(CONFIG_MACH_SUN8I_V3S) || defined(CONFIG_MACH_SUN8I_S3)
 static int sunxi_get_boot_source(void)
 {
-	return (s8)readb(SPL_ADDR + 0x6fff);
+	return (s8)readb(SPL_ADDR + 0x28);
 }
 #else
 static int sunxi_get_boot_source(void)
@@ -275,7 +275,7 @@ uint32_t sunxi_get_boot_device(void)
 		return BOOT_DEVICE_SPI;
 	}
 
-	panic("Unknown boot source %d\n", boot_source);
+	panic("Sunxi unknown boot source %d\n", boot_source);
 	return -1;		/* Never reached */
 }
 
diff --git a/configs/allwinner_s3_generic_defconfig b/configs/allwinner_s3_generic_defconfig
index 0eace0ad6a..71b625ad0b 100644
--- a/configs/allwinner_s3_generic_defconfig
+++ b/configs/allwinner_s3_generic_defconfig
@@ -34,7 +34,6 @@ CONFIG_PWM_SUNXI=y
 CONFIG_BACKLIGHT_PWM=y
 
 # USB
-CONFIG_CMD_DFU=y
 CONFIG_PHY=y
 CONFIG_PHY_SUN4I_USB=y
 CONFIG_USB_GADGET_VBUS_DRAW=250
@@ -107,3 +106,12 @@ CONFIG_CMD_SPI=y
 CONFIG_SPL_SPINAND_SUNXI_SPL_SIZE=24576
 CONFIG_SPL_SPINAND_SUNXI_UBOOT_PADDING=8192
 CONFIG_SPL_SPINAND_SUNXI_PAGESIZE=2048
+
+CONFIG_CMD_DFU=y
+CONFIG_SPL_DFU=y
+CONFIG_DFU=y
+CONFIG_DFU_OVER_USB=y
+CONFIG_DFU_TIMEOUT=y
+CONFIG_DFU_MMC=y
+CONFIG_DFU_SF=y
+CONFIG_DFU_MTD=y
diff --git a/drivers/dfu/dfu_sf.c b/drivers/dfu/dfu_sf.c
index 0fdbfae434..22ba8e9896 100644
--- a/drivers/dfu/dfu_sf.c
+++ b/drivers/dfu/dfu_sf.c
@@ -13,6 +13,13 @@
 #include <jffs2/load_kernel.h>
 #include <linux/mtd/mtd.h>
 
+#ifndef CONFIG_SF_DEFAULT_SPEED
+#define CONFIG_SF_DEFAULT_SPEED	24000000
+#endif
+#ifndef CONFIG_SF_DEFAULT_MODE
+#define CONFIG_SF_DEFAULT_MODE	0
+#endif
+
 static int dfu_get_medium_size_sf(struct dfu_entity *dfu, u64 *size)
 {
 	*size = dfu->data.sf.size;
