--- a/tornado/test/asyncio_test.py	2024-12-20 11:48:39.784499740 +0800
+++ b/tornado/test/asyncio_test.py	2024-12-20 11:46:32.559892068 +0800
@@ -46,7 +46,7 @@
         if hasattr(asyncio, 'ensure_future'):
             ensure_future = asyncio.ensure_future
         else:
-            ensure_future = asyncio.async
+            ensure_future = asyncio.create_task
 
         x = yield ensure_future(
             asyncio.get_event_loop().run_in_executor(None, lambda: 42))
